crypto/evp/p_lib.c

```diff
@@ -106,43 +126,163 @@ int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from)
         return 0;
     }
 
-    if (from->ameth && from->ameth->param_copy)
+    /*
+     * If |from| is provided, we upgrade |to| to be provided as well.
+     * This drops the legacy key from |to|.
+     * evp_pkey_upgrade_to_provider() checks if |to| is already provided,
+     * we don't need to do that here.
+     */
+    if (from->keymgmt != NULL) {
+        EVP_KEYMGMT *tmp_keymgmt = from->keymgmt;
+
+        evp_pkey_upgrade_to_provider(to, NULL, &tmp_keymgmt, NULL);
+    }
+
+    /* For purely provided keys, we just call the keymgmt utility */
+    if (to->keymgmt != NULL && from->keymgmt != NULL)
+        return evp_keymgmt_util_copy(to, (EVP_PKEY *)from,
```

Comment:
maybe one of the names for evp_keymgmt_util_copy / evp_keymgmt_copy should be updated to make it really clear what the difference is.

Reply:
That's what "util" is supposed to do.

Reply:
Using the name 'util' doesnt make it obvious what the difference is. From the name I should get some indication of what the difference is.

Reply:
Suggest something other than "util" then

Reply:
Describe to me what the difference is in 2 words or less :)

Reply:
!util == wrapper

Did that help? :wink:

Reply:
Not really - I give up ;)

Reply:
What I meant to say is that `evp_keymgmt_util_` are helper / utility functions that deal with provider side keys only, while other `evp_keymgmt_` functions are wrappers for the keymgmt downcalls to the provider.